Code quality improvements at Home Connect

Diegorro98 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

- Added types to all arguments and return values to all functions
- Defined class members and its types outside the constructor
- Improved logic at binary sensor

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Diegorro98 commented:
@joostlek These are all the things I found that I think they need quality improvements, do you know any tool or command that can help me find any other code quality issues?

joostlek commented on homeassistant/components/home_connect/__init__.py:
```diff
@@ -244,7 +244,7 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
 @Throttle(SCAN_INTERVAL)
-async def update_all_devices(hass, entry):
+async def update_all_devices(hass: HomeAssistant, entry: ConfigEntry):
```
return type?

joostlek commented on homeassistant/components/home_connect/__init__.py:
```diff
@@ -244,7 +244,7 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
 
 @Throttle(SCAN_INTERVAL)
```
We want to remove as much throttles as possible, so it would be interesting to explore the option of adding a coordinator

Diegorro98 replied:
Maybe we can create an issue to add this to the backlog, because doing this change in this PR will take time and will slow this PR and the other pending PRs.

joostlek commented on homeassistant/components/home_connect/api.py:
```diff
@@ -42,6 +43,11 @@
 class ConfigEntryAuth(homeconnect.HomeConnectAPI):
     """Provide Home Connect authentication tied to an OAuth2 based config entry."""
 
+    hass: core.HomeAssistant
+    config_entry: config_entries.ConfigEntry
+    session: config_entry_oauth2_flow.OAuth2Session
+    devices: list[dict[str, Any]]
```
You don't need to overwrite these types as these parameters are already typed

Diegorro98 replied:
devices type cannot be infered at initialziation, should I set the type at `__init__` (what it was already) or should I define it here?

joostlek commented on homeassistant/components/home_connect/api.py:
```diff
@@ -110,13 +117,17 @@ class HomeConnectDevice:
 
-    def __init__(self, hass, appliance):
+    def __init__(
+        self, hass: core.HomeAssistant, appliance: HomeConnectAppliance
```
```suggestion
        self, hass: HomeAssistant, appliance: HomeConnectAppliance
```
Please don't import core

joostlek commented:
We want our functions completely typed in general. Instance variables are usually inferred when set (unless mypy complains, then you add one). So we don't have to overwrite every instance variable with a type outside of the constructor

Diegorro98 commented on homeassistant/components/home_connect/entity.py:
```diff
@@ -17,10 +17,11 @@ class HomeConnectEntity(Entity):
     def __init__(self, device: HomeConnectDevice, desc: str) -> None:
         """Initialize the entity."""
-        self.device = device
+        self._attr_device = device
```
Should I leave it as `device` and delete the property definition?

joostlek replied:
I think that works yes. The only thing that this setup does is to make it read only to avoid `self.device = ...`, but I don't think there's any significance in doing this here

joostlek commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -148,7 +147,9 @@ async def async_update(self) -> None:
     """Switch class for Home Connect."""
 
-    def __init__(self, device, program_name):
+    program_name: str
```
Isn't this one also inferred?

Diegorro98 committed:
Code quality improvements at Home Connect

Added types to all arguments and return values to all functions

Defined class members and its types outside the constructor

Improved logic at binary sensor

Diegorro98 commented:
I noticed that in some components, the entity description tuples are placed under the classes. This might be a good idea because entity descriptions can be quite large, and having the classes at the end of the file means you have to scroll all the way down to get to the code, which is quite inconvenient and sensor. (And also specifying the type of the tuples might not be neccesary)
https://github.com/home-assistant/core/blob/9189af7e84a36e2deafb88cabad58ab231008c9f/homeassistant/components/home_connect/binary_sensor.py#L52-L65
https://github.com/home-assistant/core/blob/9189af7e84a36e2deafb88cabad58ab231008c9f/homeassistant/components/home_connect/light.py#L52-L63
https://github.com/home-assistant/core/blob/9189af7e84a36e2deafb88cabad58ab231008c9f/homeassistant/components/home_connect/sensor.py#L53-L90
https://github.com/home-assistant/core/blob/9189af7e84a36e2deafb88cabad58ab231008c9f/homeassistant/components/home_connect/switch.py#L40-L54

Diegorro98 commented:
Also, I had a question that this PR might be the place to be answered to then make the necessary changes and add them to this PR.
At `strings.json`, is there any specified order for the first level keys?
I've been setting them always alphabetically ordered, but right now in this component, `entity` object is at the end of the file, should we move it?

joostlek commented:
Usually the order in the file is

Entity definitions
Setup logic
Entity classes

Diegorro98 commented:
@joostlek regarding the `strings.json` file, does it require a particular order? I'm asking because on the dev branch, assuming alphabetical sorting is used, the `entity` object seems to be misplaced in this component.

joostlek commented:
We don't really enforce an order in the strings.json, but it usually starts with the config flow and then the rest
